stdlib/json.gr

```diff
@@ -0,0 +1,2103 @@
+   *  escapeNonASCII: false,
+   * }
+   * ```
+   *
+   * The following example have whitespaces, line breaks and control points
+   * replaced with visible characters.
+   * ```
+   * {"currency":"€","price":99.9,"currencyDescription":"EURO␡","script·unembeddable":"You·cannot·paste·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·</·with·unescaped·forward·slash"}
+   * ```
+   *
+   * @since v0.6.0
+   */
+  Compact,
+  /**
+   * Pretty and conservative formatting to maximize compatibility and
+   * embeddability of the resulting JSON.
+   *
+   * Should be safe to copy and paste directly into HTML and to be transported in
+   * plain ASCII.
+   *
+   * Roughly Equivalent to:
+   * ```
+   * Custom{
+   *  indentation: IndentWithSpaces(2),
+   *  arrayFormat: OneArrayEntryPerLine,
+   *  objectFormat: OneObjectEntryPerLine,
+   *  lineEnding: LineFeed,
+   *  finishWithNewLine: true,
+   *  escapeAllControlPoints: true,
+   *  escapeHTMLUnsafeSequences: true,
+   *  escapeNonASCII: true,
+   * }
+   * ```
+   *
+   * The following example have whitespaces, line breaks and control points
+   * replaced with visible characters.
+   * ```
+   * {↵
+   * ··"currency":·"\u20ac",↵
+   * ··"price":·99.9,↵
+   * ··"currencyDescription":·"EURO\u007f",↵
+   * ··"script·unembeddable":·"You·cannot·paste·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·<\/·with·unescaped·forward·slash"↵
+   * }
+   * ```
+   *
+   * @since v0.6.0
+   */
+  PrettyAndSafeFormat,
+  /**
+   * Compact and conservative formatting to maximize compatibility and
+   * embeddability of the resulting JSON.
+   *
+   * Should be safe to copy and paste directly into HTML and to transported in
+   * plain ASCII.
+   *
+   * Roughly Equivalent to:
+   * ```
+   * Custom{
+   *  indentation: NoIndentation,
+   *  arrayFormat: CompactArrayEntries,
+   *  objectFormat: CompactObjectEntries,
+   *  lineEnding: NoLineEnding,
+   *  finishWithNewLine: false,
+   *  escapeAllControlPoints: true,
+   *  escapeHTMLUnsafeSequences: true,
+   *  escapeNonASCII: true,
+   * }
+   * ```
+   *
+   * The following example have whitespaces, line breaks and control points
+   * replaced with visible characters.
+   * ```
+   * {"currency":"\u20ac","price":99.9,"currencyDescription":"EURO\u007f","script·unembeddable":"You·cannot·paste·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·<\/·with·unescaped·forward·slash"}
+   * ```
+   *
+   * @since v0.6.0
+   */
+  CompactAndSafeFormat,
+  /**
+   * Allows for fined grained control of the formatting output.
+   *
+   * @since v0.6.0
+   */
+  Custom{
+    indentation: IndentationFormat,
+    arrayFormat: ArrayFormat,
+    objectFormat: ObjectFormat,
+    lineEnding: LineEnding,
+    finishWithNewLine: Bool,
+    escapeAllControlPoints: Bool,
+    escapeHTMLUnsafeSequences: Bool,
+    escapeNonASCII: Bool,
+  },
+}
+
+record JsonWriterConfig {
+  format: FormattingSettings,
```

Comment:
Instead of having a separate `FormattingSettings` record, can we just use the type `FormattingChoices` and convert any friendly names to the `Custom` variant?

Reply:
I am not a big fan of this approach mainly from a type safety perspective. as we need to extract the values from the `Custom` variant inside the json writer. which would mean ensuring everytime we want to access the settings that it is a custom variant.

Alternatively we use `FormattingChoices` internally and dont convert it to a `Custom` variant but then we are splitting up the config for each option which I dont think is a good option.